### ISSUE TITLE
1st control plane machine takes ~20 mins to provision via ClusterClass 

### ISSUE THREAD

**Issue by jackfrancis** [kubernetes-sigs/cluster-api-provider-azure#2505]
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

We're seeing a very consistent repro in our `ClusterClass` test scenario where it takes roughly 20 mins to provision the first control plane machine, which fails tests (the timeout for this phase of the test run is currently configured to 20 mins).

**What did you expect to happen:**

All other test scenarios are doing fine, where the control plane comes online prior to the 20 minute timeout.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

You can see from the capz-controller-manager logs that the first reconciliation happens at :05

```
I0725 11:05:39.774324       1 azuremachine_controller.go:243] controller/azuremachine/controllers.AzureMachineReconciler.reconcileNormal "msg"="Reconciling AzureMachine" "name"="capz-e2e-0a5gl6-cc-md-0-infra-77gkc-2szx9" "namespace"="capz-e2e-0a5gl6" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureMachine" "x-ms-correlation-request-id"="aa74266d-8388-44b7-ab73-7d86428424af" 
I0725 11:05:39.789127       1 azuremachine_controller.go:266] controller/azuremachine/controllers.AzureMachineReconciler.reconcileNormal "msg"="Bootstrap data secret reference is not yet available" "name"="capz-e2e-0a5gl6-cc-md-0-infra-77gkc-2szx9" "namespace"="capz-e2e-0a5gl6" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureMachine" "x-ms-correlation-request-id"="aa74266d-8388-44b7-ab73-7d86428424af" 
```

And then it finally reconciles around 20 mins later:

```
I0725 11:24:14.676850       1 async.go:111] controller/azuremachine/async.Service.CreateResource "msg"="successfully got existing resource" "name"="capz-e2e-0a5gl6-cc-control-plane-cspcg-cbgnf" "namespace"="capz-e2e-0a5gl6" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureMachine" "x-ms-correlation-request-id"="183bf88a-8fb2-469c-868d-cf02c1b759b5" "resource"="capz-e2e-0a5gl6-cc-control-plane-cspcg-cbgnf" "resourceGroup"="capz-e2e-0a5gl6-cc-h2d54" "service"="inboundnatrules"
I0725 11:24:14.676910       1 async.go:120] controller/azuremachine/async.Service.CreateResource "msg"="resource up to date" "name"="capz-e2e-0a5gl6-cc-control-plane-cspcg-cbgnf" "namespace"="capz-e2e-0a5gl6" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureMachine" "x-ms-correlation-request-id"="183bf88a-8fb2-469c-868d-cf02c1b759b5" "resource"="capz-e2e-0a5gl6-cc-control-plane-cspcg-cbgnf" "resourceGroup"="capz-e2e-0a5gl6-cc-h2d54" "service"="inboundnatrules"
I0725 11:24:15.168473       1 async.go:80] controller/azuremachine/async.Service.processOngoingOperation "msg"="long running operation has completed" "name"="capz-e2e-0a5gl6-cc-control-plane-cspcg-cbgnf" "namespace"="capz-e2e-0a5gl6" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureMachine" "x-ms-correlation-request-id"="183bf88a-8fb2-469c-868d-cf02c1b759b5" "resource"="capz-e2e-0a5gl6-cc-control-plane-cspcg-cbgnf-nic" "service"="interfaces"
```

**Environment:**

- cluster-api-provider-azure version: main branch
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by jackfrancis** [comment]
Not sure if this is the root cause, but this does seem to be the reconciliation phase that is failing:

- https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/4be3793d7d8d07971060ebc2228cb339136556fd/controllers/azuremachine_controller.go#L264-L269
- https://github.com/kubernetes-sigs/cluster-api/blob/v1.1.4/internal/controllers/machine/machine_controller_phases.go#L219-L223

**Comment by Jont828** [comment]
TODO:
- Creating unneeded machine templates like ci-default where owner is ClusterClass. Probably shouldn't be reconciled.
- Nil pointer issue is causing a return w/ 10 min period
- MachineTemplate has strange annotation: `topology.cluster.x-k8s.io/managed-field-paths: H4sIAAAAAAAA/6pWKknNLchJLElVsqpWKi5ITQbRKYkliS6ZxdnFSlbVtTpKmSmpeSWZJZUQXn4xSAqkLDkxOSMzLz2ksiAVIpWSWZwdnFmV6u4EUwqTq9VRKi7OCChNyslM9k6FmlSWC1IMlq6tBQQAAP//K5jpp4oAAAA=`
- MachineTemplate has label set to empty string: `topology.cluster.x-k8s.io/owned: ""`
- reconcileExternalReference in KubeadmControlPlaneReconciler sets owner ref to cluster instead of upserting/appending. This could be overwriting a clusterclass owner, but need to verify.

**Comment by jackfrancis** [comment]
> Nil pointer issue is causing a return w/ 10 min period, see https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2489

Great find @Jont828, I confirmed that the above change is when this ClusterClass E2E symptom started happening.

Based on your other comments, it seems possible that the ClusterClass template we're using to test E2E is not entirely hygienic? Because we're not seeing these symptoms elsewhere I would assume the azurejson nil pointer check is desirable, and we just need to clean up the way we're deploying ClusterClass-backed clusters in E2E. (just an assumption at this point though!)

**Comment by Jont828** [comment]
@jackfrancis There's a lot to break down and I'll get rid of that TODO comment. To answer your question, yes we are setting up the templates incorrectly since we are creating unneeded machine templates. However, I believe that's a separate issue. 

For the ClusterClass test, one key difference is that normal Clusters have an infraRef in their spec from the moment it gets created, but for Clusters created from a ClusterClass, it doesn't have one and instead waits for one to be created based on an infraTemplate. I believe the reason for the timeout here is that we return w/o an error when the infraRef isn't present, causing us to wait 10 minutes before we can check again. A possible solution would be to have the AzureJSONMachineController watch the Cluster object and run again when the infraRef is added so we avoid the downtime.
